### PR TITLE
feat: Use Session Manager to save Session

### LINKED ISSUE #437: Save session as a file
**Is your feature request related to a problem? Please describe.**

The current way of handling session doesn't work in all cases

For example, it cannot restore the previous session if it is a force-ful shut down

**Describe the solution you'd like**

Like the title, save session as a file

### PR DESCRIPTION
**Pull request by coder3101**
## Description
Use Session manager (new API) to save and restore and periodically save sessions instead of using Settings Manager.

## Related Issue
Fixes #437 
 
## Motivation and Context
In hard power-down of Computer, the Status is lost if unsaved. With the help of this new API, we can periodically save the session to file. This also makes Session independent of SettingsManager. So Sessions are saved in a separate JSON File and could be restored on any CP Editor on any OS.

## How Has This Been Tested?
Windows 10

## Screenshots (if appropriate)
image

## Type of changes

- [ ] Bug fix (changes which fix an issue)
- [x] New feature (changes which add functionality)
- [ ] Documentation (changes which modify the documentation only)
- [ ] Style change (changes which do not affect the meaning of the code: code formatting, etc.)
- [ ] Refactor (changes which affect the meaning of the code but neither fix a bug nor add a feature)
- [ ] Performance improve (changes which improve performance)
- [ ] Test (changes which add tests)
- [ ] Build (Changes that affect the build system or external dependencies)
- [ ] CI (changes to CI configuration files and scripts)
- [ ] Chore (changes which do not belong to any type above)
- [ ] Revert (revert previous changes)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/cpeditor/cpeditor/blob/master/CONTRIBUTING.md) document.
- [x] I have tested these changes locally, and this fixes the bug/the new feature behaves as the expectation.
- [x] The settings file in the old version can be used in the new version after this change.
- [x] These changes only fix a single bug/introduces a single feature. (Otherwise, open multiple Pull Requests instead, unless these bugs/features are closely related.)
- [x] The commit messages are clear and detailed. (Otherwise, use `git reset` and commit again, or use `git rebase -i` and `git commit --amend` to modify the commit messages.)
- [x] These changes don't remove an existing feature. (Otherwise, add an option to disable this feature instead, unless it's necessary to remove this feature.)
- [x] If there are changes of the text displayed in the UI, I have wrapped them in `tr()` or `QCoreApplication::translate()`.
- [ ] I have documented these changes in [CHANGELOG.md](https://github.com/cpeditor/cpeditor/blob/master/doc/CHANGELOG.md), or these changes are not notable.

### CODE REVIEW

**Comment by coder3101** [comment]
Checklist:
- [ ] Add to changelog
- [ ] Add Settings for Periodically saving Editor Session. (Session Save Interval, Set to 0 means Do not save periodically otherwise save session after every `n` ms, `n` is value of Session Save Interval.)
- [ ] Add translation
- [ ] Resolve Conflicts

**Comment by coder3101** [comment]
@ouuan Please add the translations, Settings and Changelogs and let me know for a quick review before a merge!

**Comment by ouuan** [comment]
What's `restoreSession(false)` used for?

**Comment by coder3101** [comment]
> What's `restoreSession(false)` used for?

No use, I will remove it now.

File: src/Core/SessionManager.hpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+namespace Core
+{
+class SessionManager : public QObject
+{
+    Q_OBJECT
+
+  public:
+    static void initiate(AppWindow *appwindow);
+
+    // This function should be called before initiate(). The call to initiate() deletes the last session file.
```
Comment (coder3101) [inline src/Core/SessionManager.hpp:37]: The call to `updateSession()` deletes the last session file not the `initiate()`

**Comment by neko-para** [comment]
I'm wondering if we should add a class or a function for path replacing?

**Comment by coder3101** [comment]
> I'm wondering if we should add a class or a function for path replacing?

Some function in FileUtil will be nice!

**Comment by ouuan** [comment]
I'm doing code refactoring.

**Comment by ouuan** [comment]
Please wait until my next commit.

**Comment by neko-para** [comment]
That's a sad story. I'm doing as well. I'll wait your commit and rebase mine.

**Comment by ouuan** [comment]
It has been not working at least since 791cbea, so I haven't tested it yet.

**Comment by coder3101** [comment]
> 
> 
> It has been not working at least since [791cbea](https://github.com/cpeditor/cpeditor/commit/791cbea6e25866a35f2efb5c0578fd36de843387), so I haven't tested it yet.

It is working on my machine. The HEAD of this branch!

**Comment by neko-para** [comment]
I'm doing refactoring too.

**Comment by coder3101** [comment]
You can approve and merge this to `master`

**Comment by neko-para** [comment]
Wait, I've found some text miss translation

**Comment by neko-para** [comment]
@ouuan Plz check the new translation

**Comment by ouuan** [comment]
That's not related to this PR, we'd better open a new PR to fix it.

File: src/Settings/settings.json [hunk 2]
```diff
@@ -241,6 +241,9 @@
         "type": "int",
         "default": 1000,
         "param": "QVariantList {100, 36000000}",
+        "depends": [
+            "Auto Save"
```
Comment (coder3101) [inline src/Settings/settings.json:245]: an exception may be thrown in function `main` which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
```cpp
int main()
    ^
```
Reported as error by clang-tidy.

**Comment by coder3101** [comment]
Do not take the above review seriously. I was just testing a script, I wrote for automatically review a PR by clang-tidy.

**Pull request merged**